src/Components/Components/src/Lifetime/PersistentComponentState.cs

```diff
@@ -105,27 +93,31 @@ public bool TryTakePersistedState(string key, [MaybeNullWhen(false)] out byte[]?
         }
 
         /// <summary>
-        /// Persists the serialized state <paramref name="value"/> for the given <paramref name="key"/>.
+        /// Persists the serialized state <paramref name="valueWriter"/> for the given <paramref name="key"/>.
         /// </summary>
         /// <param name="key">The key to use to persist the state.</param>
-        /// <param name="value">The state to persist.</param>
-        public void PersistState(string key, byte[] value)
+        /// <param name="valueWriter">The state to persist.</param>
+        public void Persist(string key, Action<IBufferWriter<byte>> valueWriter)
         {
             if (key is null)
             {
                 throw new ArgumentNullException(nameof(key));
             }
 
-            if (value is null)
+            if (valueWriter is null)
             {
-                throw new ArgumentNullException(nameof(value));
+                throw new ArgumentNullException(nameof(valueWriter));
             }
 
             if (_currentState.ContainsKey(key))
             {
                 throw new ArgumentException($"There is already a persisted object under the same key '{key}'");
             }
-            _currentState.Add(key, value);
+
+            var pipe = new Pipe();
```

Comment:
What's consuming the other side of the pipe? Is this just to get an IBufferWriter implementation? If yes, I don't think you need a pipe here. You want something like this https://source.dot.net/#System.Text.Json/PooledByteBufferWriter.cs,d0112696348c0f6d,references

Reply:
Yep, that seems like an improvement. Let me see if I can replace the current implementation with that one.

Reply:
One issue with this implementation is that it uses a single buffer. Something like [this](https://github.com/dotnet/aspnetcore/blob/main/src/SignalR/common/Shared/MemoryBufferWriter.cs) might be better if we need to support bigger payloads.

Reply:
Yep, I have some perf numbers from a benchmark that I've created for this (if we do perf work, we should have benchmarks)